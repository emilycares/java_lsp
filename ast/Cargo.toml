[package]
name = "ast"
version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
documentation.workspace = true
readme.workspace = true
license.workspace = true

[dependencies]
phf.workspace = true
smol_str.workspace = true

[dev-dependencies]
insta.workspace = true
criterion.workspace = true

[[bench]]
name = "base"
harness = false
