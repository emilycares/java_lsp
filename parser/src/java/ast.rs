use super::{
    error::{AstError, ExpectedToken, InvalidToken, assert_token},
    lexer::{PositionToken, Token},
};

#[derive(Debug, PartialEq)]
pub struct AstFile {
    package: String,
    imports: Vec<String>,
    thing: AstThing,
}
#[derive(Debug, PartialEq)]
pub enum AstAvailability {
    Public,
    Private,
    Protected,
}

#[derive(Debug, PartialEq)]
pub struct AstClass {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub struct AstInterface {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub enum AstThing {
    Class(AstClass),
    Interface(AstInterface),
    None,
}

pub fn parse_file(tokens: &[PositionToken], pos: usize) -> Result<AstFile, AstError> {
    let (package_name, pos) = parse_package(tokens, pos)?;
    let (imports, pos) = parse_imports(tokens, pos)?;
    let (thing, _pos) = parse_thing(tokens, pos)?;

    Ok(AstFile {
        package: package_name,
        imports,
        thing,
    })
}

/// package ch.emilycares;
fn parse_package(tokens: &[PositionToken], pos: usize) -> Result<(String, usize), AstError> {
    let mut next_pos = pos;
    let mut iter = tokens.iter();
    assert_token(&mut iter, Token::Package)?;
    let mut package_name = String::new();
    for t in iter {
        next_pos += 1;
        match &t.token {
            Token::Identifier(id) => package_name.push_str(id),
            Token::Dot => package_name.push('.'),
            Token::Semicolon => break,
            _ => continue,
        }
    }
    Ok((package_name, next_pos))
}

///  import java.io.IOException;
///  import java.net.Socket;
fn parse_imports(tokens: &[PositionToken], pos: usize) -> Result<(Vec<String>, usize), AstError> {
    let mut next_pos = pos + 1;
    let mut imports = vec![];

    while let Ok((import, new_pos)) = parse_import(tokens, next_pos) {
        next_pos = new_pos;
        imports.push(import);
    }

    Ok((imports, next_pos))
}

///  import java.io.IOException;
fn parse_import(tokens: &[PositionToken], pos: usize) -> Result<(String, usize), AstError> {
    let mut iter = tokens.iter().skip(pos);
    assert_token(&mut iter, Token::Import)?;

    let mut pos = pos;
    let mut import_name = String::new();
    for t in iter {
        pos += 1;
        match &t.token {
            Token::Identifier(id) => import_name.push_str(id),
            Token::Dot => import_name.push('.'),
            Token::Star => import_name.push('*'),
            Token::Semicolon => {
                pos += 1;
                break;
            }
            _ => continue,
        }
    }
    Ok((import_name, pos))
}

///  public class Everything { ...
///  public interface Constants { ...
fn parse_thing(tokens: &[PositionToken], pos: usize) -> Result<(AstThing, usize), AstError> {
    let mut pos = pos;
    let mut iter = tokens.iter().skip(pos);
    pos += 1;
    let avaliability = parse_avaliability(iter.next())?;
    match iter.next() {
        Some(t) => match t {
            PositionToken {
                token: Token::Class,
                line: _,
                col: _,
            } => Ok((AstThing::Class(AstClass { avaliability }), pos)),
            PositionToken {
                token: Token::Interface,
                line: _,
                col: _,
            } => Ok((AstThing::Interface(AstInterface { avaliability }), pos)),
            found => Err(AstError::ExpectedToken(ExpectedToken::from(
                found,
                Token::Class,
            ))),
        },
        None => Err(AstError::UnexpectedEOF),
    }
}

fn parse_avaliability(token: Option<&PositionToken>) -> Result<AstAvailability, AstError> {
    let Some(token) = token else {
        return Err(AstError::UnexpectedEOF);
    };
    match token {
        PositionToken {
            token: Token::Public,
            line: _,
            col: _,
        } => Ok(AstAvailability::Public),
        PositionToken {
            token: Token::Private,
            line: _,
            col: _,
        } => Ok(AstAvailability::Private),
        PositionToken {
            token: Token::Protected,
            line: _,
            col: _,
        } => Ok(AstAvailability::Protected),
        e => Err(AstError::InvalidAvailability(InvalidToken::from(e))),
    }
}

#[cfg(test)]
pub mod tests {
    use pretty_assertions::assert_eq;

    use crate::java::error::PrintErr;
    use crate::java::{
        ast::{AstAvailability, AstClass, AstFile, AstInterface, AstThing, parse_file},
        lexer,
    };

    #[test]
    fn everything() {
        let content = include_str!("../../test/Everything.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec![],
                thing: AstThing::Class(AstClass {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }

    #[test]
    fn superee() {
        let content = include_str!("../../test/Super.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec!["java.io.IOException".to_string()],
                thing: AstThing::Class(AstClass {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }

    #[test]
    fn constants() {
        let content = include_str!("../../test/Constants.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec![
                    "jdk.net.Sockets".to_string(),
                    "java.io.IOException".to_string(),
                    "java.net.Socket".to_string()
                ],
                thing: AstThing::Interface(AstInterface {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }
}
