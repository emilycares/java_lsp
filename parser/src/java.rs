use std::str::Utf8Error;

use ast::{
    lexer,
    types::{
        AstAnnotationField, AstClassMethod, AstClassVariable, AstExtends, AstFile, AstImports,
        AstInterfaceConstant, AstInterfaceMethod, AstInterfaceMethodDefault, AstJTypeKind,
        AstThing, AstTypeParameters,
    },
};

use crate::{
    dto::{self, Access, Field, ImportUnit, Method},
    loader::SourceDestination,
};

#[derive(Debug)]
pub enum ParseJavaError {
    Utf8(Utf8Error),
    Class(dto::ClassError),
    Io(std::io::Error),
    UnknownJType(String, String),
    UnknownWildcard(String),
    Ast(ast::error::AstError),
    Lexer(ast::lexer::LexerError),
}
pub fn load_java(
    bytes: &[u8],
    source: SourceDestination,
) -> Result<crate::dto::Class, ParseJavaError> {
    let str = str::from_utf8(bytes).map_err(ParseJavaError::Utf8)?;
    let tokens = lexer::lex(str).map_err(ParseJavaError::Lexer)?;
    let parsed = ast::parse_file(&tokens).map_err(ParseJavaError::Ast)?;
    load_java_tree(parsed, source)
}

pub fn load_java_tree(
    ast: AstFile,
    source: SourceDestination,
) -> Result<crate::dto::Class, ParseJavaError> {
    let thing = ast.thing;
    let mut methods: Vec<Method> = vec![];
    let mut fields: Vec<Field> = vec![];
    let class_path_base: String = ast.package.into();
    let name;
    let mut super_class = dto::SuperClass::None;
    let mut super_interfaces = vec![];
    let imports: Vec<ImportUnit> = ast.imports.imports.iter().map(|i| i.into()).collect();
    match thing {
        AstThing::Class(class) => {
            name = class.name.into();
            methods.extend(class.methods.iter().map(convert_class_method));
            fields.extend(class.variables.iter().map(convert_class_field));
            super_class = match class.superclass {
                ast::types::AstSuperClass::None => dto::SuperClass::None,
                ast::types::AstSuperClass::Name(ast_identifier) => {
                    dto::SuperClass::Name(ast_identifier.into())
                }
                ast::types::AstSuperClass::ClassPath(ast_identifier) => {
                    dto::SuperClass::ClassPath(ast_identifier.into())
                }
            };
        }
        AstThing::Enumeration(enumeration) => {
            name = enumeration.name.into();
            methods.extend(enumeration.methods.iter().map(convert_class_method));
            fields.extend(enumeration.variables.iter().map(convert_class_field));
        }
        AstThing::Interface(interface) => {
            name = interface.name.into();
            if let Some(ext) = interface.extends {
                super_interfaces.extend(fun_name(&ext, &imports));
            }
            methods.extend(interface.methods.iter().map(convert_interface_method));
            methods.extend(
                interface
                    .default_methods
                    .iter()
                    .map(convert_interface_default_method),
            );
            fields.extend(interface.constants.iter().map(convert_interface_constant));
        }
        AstThing::Annotation(annotation) => {
            name = annotation.name.into();
            fields.extend(annotation.fields.iter().map(convert_annotation_field));
        }
    }
    let source = match source {
        SourceDestination::RelativeInFolder(e) => {
            format!("{}/{}/{}.java", e, &class_path_base.replace(".", "/"), name)
        }
        SourceDestination::Here(e) => e.replace("\\", "/"),
        SourceDestination::None => "".to_string(),
    };

    Ok(dto::Class {
        source,
        class_path: format!("{class_path_base}.{name}"),
        access: vec![],
        super_class,
        super_interfaces,
        imports: convert_imports(ast.imports, class_path_base),
        name,
        methods,
        fields,
    })
}

fn fun_name(ext: &AstExtends, imports: &[ImportUnit]) -> impl Iterator<Item = dto::SuperClass> {
    ext.parameters.iter().filter_map(|i| {
        if let AstJTypeKind::Class(c) = &i.value {
            return match imports
                .iter()
                .find_map(|i| i.get_imported_class_package(&c.value))
            {
                Some(class_path) => Some(dto::SuperClass::ClassPath(class_path)),
                None => Some(dto::SuperClass::Name(c.into())),
            };
        }
        None
    })
}

fn convert_imports(imports: AstImports, package: String) -> Vec<ImportUnit> {
    let mut out = vec![ImportUnit::Package(package)];
    out.extend(imports.imports.iter().map(|i| i.into()));
    out
}

fn convert_class_method(m: &AstClassMethod) -> Method {
    let mut access = vec![];
    if m.header.stat {
        access.push(dto::Access::Static);
    }
    let avaliability = Access::from(&m.header.avaliability, Access::Protected);
    access.push(avaliability);
    let parameters = m
        .header
        .parameters
        .parameters
        .iter()
        .map(|p| dto::Parameter {
            name: Some((&p.name).into()),
            jtype: check_type_parameters(&p.jtype, &m.header.type_parameters),
        })
        .collect();
    let mut throws = vec![];
    if let Some(t) = &m.header.throws {
        throws = t.parameters.iter().map(|i| i.into()).collect();
    }
    Method {
        access,
        name: (&m.header.name).into(),
        parameters,
        throws,
        ret: check_type_parameters(&m.header.jtype, &m.header.type_parameters),
        source: None,
    }
}
fn convert_interface_method(m: &AstInterfaceMethod) -> Method {
    let mut access = vec![];
    if m.header.stat {
        access.push(dto::Access::Static);
    }
    let avaliability = Access::from(&m.header.avaliability, Access::Public);
    access.push(avaliability);
    let parameters = m
        .header
        .parameters
        .parameters
        .iter()
        .map(|p| dto::Parameter {
            name: Some((&p.name).into()),
            jtype: check_type_parameters(&p.jtype, &m.header.type_parameters),
        })
        .collect();
    let mut throws = vec![];
    if let Some(t) = &m.header.throws {
        throws = t
            .parameters
            .iter()
            .map(|i| check_type_parameters(i, &m.header.type_parameters))
            .collect();
    }
    Method {
        access,
        name: (&m.header.name).into(),
        parameters,
        throws,
        ret: check_type_parameters(&m.header.jtype, &m.header.type_parameters),
        source: None,
    }
}
fn convert_interface_default_method(m: &AstInterfaceMethodDefault) -> Method {
    let mut access = vec![];
    if m.header.stat {
        access.push(dto::Access::Static);
    }
    let avaliability = Access::from(&m.header.avaliability, Access::Public);
    access.push(avaliability);
    let parameters = m
        .header
        .parameters
        .parameters
        .iter()
        .map(|p| dto::Parameter {
            name: Some((&p.name).into()),
            jtype: check_type_parameters(&p.jtype, &m.header.type_parameters),
        })
        .collect();
    let mut throws = vec![];
    if let Some(t) = &m.header.throws {
        throws = t
            .parameters
            .iter()
            .map(|i| check_type_parameters(i, &m.header.type_parameters))
            .collect();
    }
    Method {
        access,
        name: (&m.header.name).into(),
        parameters,
        throws,
        ret: check_type_parameters(&m.header.jtype, &m.header.type_parameters),
        source: None,
    }
}

fn convert_interface_constant(c: &AstInterfaceConstant) -> dto::Field {
    dto::Field {
        access: vec![Access::from(&c.avaliability, Access::Public)],
        name: (&c.name).into(),
        jtype: (&c.jtype).into(),
        source: None,
    }
}
fn convert_annotation_field(c: &AstAnnotationField) -> dto::Field {
    dto::Field {
        access: vec![],
        name: (&c.name).into(),
        jtype: (&c.jtype).into(),
        source: None,
    }
}
fn convert_class_field(c: &AstClassVariable) -> dto::Field {
    dto::Field {
        access: vec![Access::from(&c.avaliability, Access::Public)],
        name: (&c.name).into(),
        jtype: (&c.jtype).into(),
        source: None,
    }
}

fn check_type_parameters(
    jtype: &ast::types::AstJType,
    type_parameters: &Option<AstTypeParameters>,
) -> dto::JType {
    let jtype: dto::JType = jtype.into();
    let Some(type_parameters) = type_parameters else {
        return jtype;
    };

    if let dto::JType::Class(ref p) = jtype {
        if type_parameters.parameters.iter().any(|i| i.value == p) {
            return dto::JType::Parameter(p.to_owned());
        }
    }
    if let dto::JType::Generic(name, params) = jtype {
        let params = params
            .iter()
            .map(|i| {
                if let dto::JType::Class(p) = i {
                    if type_parameters.parameters.iter().any(|i| i.value == p) {
                        return dto::JType::Parameter(p.to_owned());
                    }
                }
                i.clone()
            })
            .collect();
        return dto::JType::Generic(name, params);
    }

    jtype
}

#[cfg(test)]
pub mod tests {
    use crate::loader::SourceDestination;

    use super::load_java;

    #[test]
    fn jtype_recognition() {
        let result = load_java(
            include_bytes!("../test/Types.java"),
            SourceDestination::Here("/path/to/source/Test.java".to_string()),
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn super_class() {
        let content = r#"
package a.test;
public class Test extends a { }
        "#;
        let result = load_java(
            content.as_bytes(),
            SourceDestination::Here("/path/to/source/Test.java".to_string()),
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn generic_type_declare() {
        let content = r#"
package a.test;
public class Test {
  public static <T> int add(Collection<T> list, T item){}
}
        "#;
        let result = load_java(
            content.as_bytes(),
            SourceDestination::Here("/path/to/source/Test.java".to_string()),
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn thrower() {
        let content = include_str!("../test/Thrower.java");
        let result = load_java(
            content.as_bytes(),
            SourceDestination::Here("/path/to/source/Thrower.java".to_string()),
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn interface_constants() {
        let result = load_java(
            include_bytes!("../test/Constants.java"),
            SourceDestination::RelativeInFolder("/path/to/source".to_string()),
        );

        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn interface_base() {
        let result = load_java(
            include_bytes!("../test/InterfaceBase.java"),
            SourceDestination::RelativeInFolder("/path/to/source".to_string()),
        );

        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn jenum() {
        let result = load_java(
            include_bytes!("../test/Variants.java"),
            SourceDestination::RelativeInFolder("/path/to/source".to_string()),
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn jannotation() {
        let result = load_java(
            include_bytes!("../test/Annotation.java"),
            SourceDestination::RelativeInFolder("/path/to/source".to_string()),
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn everything() {
        let result = load_java(
            include_bytes!("../test/Everything.java"),
            SourceDestination::None,
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }

    #[test]
    fn int() {
        let src = r#"
package a.test;

import jakarta.inject.Inject;
import jakarta.ws.rs.GET;

import jakarta.ws.rs.Path;

public class Test {
}
 "#;
        let result = load_java(
            src.as_bytes(),
            SourceDestination::RelativeInFolder("/path/to/source".to_string()),
        );

        insta::assert_debug_snapshot!(result.unwrap());
    }
    #[test]
    fn super_interfaces() {
        let result = load_java(
            include_bytes!("../test/SuperInterface.java"),
            SourceDestination::None,
        );
        insta::assert_debug_snapshot!(result.unwrap());
    }
}
