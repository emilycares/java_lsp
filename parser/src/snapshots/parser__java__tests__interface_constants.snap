---
source: parser/src/java.rs
expression: result.unwrap()
---
Class {
    class_path: "ch.emilycares.Constants",
    source: "/path/to/source/ch/emilycares/Constants.java",
    access: [],
    imports: [
        Package(
            "ch.emilycares",
        ),
        Class(
            "jdk.net.Sockets",
        ),
        Class(
            "java.io.IOException",
        ),
        Class(
            "java.net.Socket",
        ),
    ],
    name: "Constants",
    methods: [
        Method {
            access: [
                Public,
            ],
            name: "display",
            parameters: [],
            throws: [],
            ret: Void,
            source: None,
        },
        Method {
            access: [
                Public,
            ],
            name: "createSocket",
            parameters: [
                Parameter {
                    name: Some(
                        "hostname",
                    ),
                    jtype: Class(
                        "String",
                    ),
                },
                Parameter {
                    name: Some(
                        "port",
                    ),
                    jtype: Int,
                },
            ],
            throws: [
                Class(
                    "IOException",
                ),
            ],
            ret: Class(
                "Socket",
            ),
            source: None,
        },
    ],
    fields: [
        Field {
            access: [
                Public,
            ],
            name: "CONSTANT_A",
            jtype: Class(
                "String",
            ),
            source: None,
        },
        Field {
            access: [
                Public,
            ],
            name: "CONSTANT_B",
            jtype: Class(
                "String",
            ),
            source: None,
        },
        Field {
            access: [
                Public,
            ],
            name: "CONSTANT_C",
            jtype: Class(
                "String",
            ),
            source: None,
        },
    ],
    super_class: None,
    super_interfaces: [],
}
